ocaml/xapi/xapi_network_sriov_helpers.ml

```diff
@@ -124,3 +125,103 @@ let assert_sriov_pif_compatible_with_network ~__context ~pif ~network =
+
+let get_pcis_from_network ~__context ~network =
+  Db.Network.get_PIFs ~__context ~self:network
+  |> Listext.List.filter_map (fun pif -> get_underlying_pif ~__context ~pif)
+  |> List.map (fun pif -> Db.PIF.get_PCI ~__context ~self:pif)
+
+let get_remaining_on_network ~__context ~network =
+  let open Xapi_stdext_monadic in
+  try
+    let pcis = get_pcis_from_network ~__context ~network in
+    let capacity = List.fold_left
+        (fun acc pci ->
+            let remaining_capacity = Xapi_pci.get_idle_vf_nums ~__context ~self:pci |> Int64.of_int in
+            Int64.add acc remaining_capacity
+        ) 0L pcis
+    in
+    if capacity > 0L
+    then Either.Right capacity
+    else Either.Left (Api_errors.Server_error(Api_errors.network_sriov_insufficient_vfs, []))
+  with e -> Either.Left e
+
+let host_sriov_capable ~__context ~host ~network =
+  (*at most one pif in network on localhost*)
+  let local_pifs = Xapi_network_attach_helpers.get_local_pifs ~__context ~network ~host in
+  match local_pifs with
+  | local_pif :: _ ->
+    begin match get_underlying_pif ~__context ~pif:local_pif with
+      | Some underlying_pif ->
+        let pci = Db.PIF.get_PCI ~__context ~self:underlying_pif in
+        if Xapi_pci.get_idle_vf_nums ~__context ~self:pci > 0 then true else false
+      | None -> false
+    end
+  | [] -> false
+
+let group_hosts_by_best_sriov ~__context ~network =
+  let pcis = get_pcis_from_network __context network in
+  let can_allocate_vf pci = Xapi_pci.get_idle_vf_nums ~__context ~self:pci |> Int64.of_int > 0L in
+  let valid_pcis = List.filter can_allocate_vf pcis in
+  (* group hosts by breadth-first algorithm *)
+  Helpers.group_by `descending
+    (fun host ->
+        let pci = List.find (fun pci -> host = Db.PCI.get_host ~__context ~self:pci) valid_pcis in
+        Xapi_pci.get_idle_vf_nums ~__context ~self:pci |> Int64.of_int
+    ) (List.map (fun pci -> Db.PCI.get_host ~__context ~self:pci) valid_pcis |> Listext.List.setify)
+
+let assert_capacity_for_vf ~__context ~pci =
+  match Xapi_pci.get_idle_vf_nums ~__context ~self:pci with
+  | 0 -> raise (Api_errors.Server_error(Api_errors.network_sriov_insufficient_vfs, []))
+  | _ -> ()
+
+let choose_pf_for_vf ~__context ~host ~vif =
+  let network = Db.VIF.get_network ~__context ~self:vif in
+  (* at most one pif within one network on localhost*)
+  let host_pcis = get_pcis_from_network __context network |>
+                  List.filter (fun pci -> Db.PCI.get_host ~__context ~self:pci = host) in
+  match host_pcis with
+  | pci :: t -> assert_capacity_for_vf ~__context ~pci; pci
+  | [] -> raise (Api_errors.Server_error(Api_errors.vm_requires_vf_pci_for_sriov_vif, [Ref.string_of vif]))
+
+(* If exn happens during vifs reservation ,reserved vfs will be cleared*)
+let reserve_sriov_vfs ~__context ~host ~vm =
+  let sriov_networks = get_sriov_networks_from_vm __context vm in
+  let sriov_vifs = Db.VM.get_VIFs ~__context ~self:vm
```

Comment:
This is a little inefficient, since `get_sriov_networks_from_vm` calls `get_VIFs` as well. There are a lot of helper function in this module that don't actually help very much in terms of efficiency and clarity...